server/games/game_results.py

```diff
@@ -0,0 +1,136 @@
+from enum import Enum
+from collections import Counter
+from collections.abc import Mapping
+from server.decorators import with_logger
+from typing import NamedTuple
+
+
+class GameOutcome(Enum):
+    VICTORY = 'victory'
+    DEFEAT = 'defeat'
+    DRAW = 'draw'
+    MUTUAL_DRAW = 'mutual_draw'
+    UNKNOWN = 'unknown'
+
+    @classmethod
+    def from_message(cls, msg: str):
+        try:
+            return cls(msg.lower())
+        except ValueError:
+            return cls.UNKNOWN
+
+
+class GameResult(NamedTuple):
+    """
+    These are sent from each player's FA when they quit the game. 'Score'
+    depends on the number of ACUs killed, whether the player died, maybe other
+    factors.
+    """
+    reporter: int
+    army: int
+    outcome: GameOutcome
+    score: int
+
+
+@with_logger
+class GameResults(Mapping):
+    """
+    Collects all results from a single game. Allows to determine results for an
+    army and game as a whole. Supports a dict-like access to lists of results
+    for each army, but don't modify these.
+    """
+    def __init__(self):
+        Mapping.__init__(self)
+        self._back = {}
+
+    def __getitem__(self, key: int):
+        return self._back[key]
+
+    def __iter__(self):
+        return iter(self._back)
+
+    def __len__(self):
+        return len(self._back)
+
+    def add(self, result: GameResult):
+        army_results = self._back.setdefault(result.army, [])
+        army_results.append(result)
+
+    def is_mutually_agreed_draw(self, player_armies):
+        # Can't tell if we have no results
+        if not self:
+            return False
+        # Everyone has to agree to a mutual draw
+        for army in player_armies:
+            if army not in self:
+                continue
+            if any(r.outcome is not GameOutcome.MUTUAL_DRAW
+                   for r in self[army]):
+                return False
+        return True
+
+    def outcome(self, army: int) -> GameOutcome:
+        """
+        Determines what the game outcome was for a given army. Returns the
+        outcome all players agree on, excluding players that reported an
+        unknown outcome. Reports unknown outcome if players disagree.
+        """
+        if army not in self:
+            return GameOutcome.UNKNOWN
+
+        outcomes = set(r.outcome for r in self[army]
+                       if r.outcome is not GameOutcome.UNKNOWN)
+
+        if len(outcomes) == 1:
+            return outcomes.pop()
+        else:
+            return GameOutcome.UNKNOWN
+
+    def score(self, army: int, game_id=None):
+        """
+        Pick and return most frequently reported score for an army. If multiple
+        scores are most frequent, pick the largest one.
+        """
+        if army not in self:
```

Comment:
I think some pieces of the code actually expect this to raise a `KeyError`. Have a look at:
https://github.com/FAForever/server/blob/6b9aec7f04698da4b0843c2a26a82705c5faf503/server/games/game.py#L515
and
https://github.com/FAForever/server/blob/6b9aec7f04698da4b0843c2a26a82705c5faf503/server/games/game.py#L897
It even mentions it in the docstring
https://github.com/FAForever/server/blob/6b9aec7f04698da4b0843c2a26a82705c5faf503/server/games/game.py#L841

We need some tests for `persist_results` because I don't think it ever writes `-1` into the database any more. As a matter of fact that code path is not covered by the tests: https://coveralls.io/builds/25850926/source?filename=server%2Fgames%2Fgame.py#L449
which is probably why I missed this when I rewrote that function in the first place

Reply:
I think I won't restore these KeyErrors and will just fix these catch blocks - we have good neutral values (`0`, and `GameOutcome.UNKNOWN`) for when an army isn't there. It also makes sense in that it's not an exceptional condition - we might have not received any results for a given army yet, so we need a sensible value for an army with no results.

Reply:
Yea I agree with that approach. Like I mentioned, the code used to write `-1` for unknown scores, although I'm not convinced that this is unambiguous since negative scores are possible. Is it possible to have a score of 0?